test/test_nn.py

```diff
@@ -9721,6 +9721,48 @@ def _test_dropout(self, cls, device, input, memory_format=torch.contiguous_forma
         module.__repr__()
         str(module)
 
+    def _test_dropout_discontiguous(self, cls, device, memory_format=torch.contiguous_format):
+        # In this test, we verify that dropout preserves the layout and data for different memory formats.
+        # We check whether, we get same values for the output of dropout, when the probability
+        # of dropout is 0 or very close to 0.
+        # Reference: https://github.com/pytorch/pytorch/issues/47176
+        close_to_zero_p = 0.0000000001  # Should be almost zero but not zero, as for p=0 different path is taken
+        for p in [0, close_to_zero_p]:
+            inp = torch.arange(2 * 3 * 3 * 3).reshape(2, 3, 3, 3)
+            inp_discontiguous = torch.empty(2, 3, 3, 6, device=device, memory_format=memory_format)[..., ::2]
+            inp_discontiguous.copy_(inp)
+            mod = cls(p=p)
+            out = mod(inp_discontiguous)
+            self.assertEqual(inp_discontiguous, out)
+
+    def _test_dropout_mean_check(self, cls, device):
+        def _test_dropout_mean(inp, out):
+            self.assertNotEqual(inp, out)
+            self.assertEqual(inp.mean(), out.mean(), rtol=3, atol=1)
```

Comment:
this also does not test that layout was preserved. Also, rtol=3 seems way too large. Maybe try initializing tensors with `ones()`, then tolerances can be small?

Reply:
Makes sense. Will try that. Thanks!